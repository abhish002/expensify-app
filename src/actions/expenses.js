import database from '../firebase/firebase';

// Add Expense Action
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

// Dispatches Add Expense as a function
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const { description='', note='', amount=0, createdAt=0 } = expenseData;
        const expense = { description, note, amount, createdAt };

        return database.ref('expenses').push(expense).then(ref => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    }
}

// Remove Expense Action
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

// Async action to remove an expense
export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        const ref = database.ref(`expenses/${id}`);
        return ref.remove().then(() => {
            dispatch(removeExpense({ id }))
        });
    }
}

// Edit Expense Action
export const editExpense = (id, updates ) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

// Async action to edit expense
export const startEditExpense = (id, updates) => {
    return (dispatch) => {        
        return database.ref(`expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates))
        });
    }
}

// SET_EXPENSES
export const setExpenses= (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

// async action to call and set expenses
export const startSetExpenses = () => {
    return((dispatch) => {
        const expenses = [];
        return database.ref('expenses').once('value').then((snapshot) => {            
            snapshot.forEach(expense => {
                expenses.push({
                    id: expense.key,
                    ...expense.val()
                });
            });
            dispatch(setExpenses(expenses));
        });
    });
}