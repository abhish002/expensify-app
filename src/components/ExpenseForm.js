import React from 'react'
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props);
        
        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            isCalendarFocused: false,
            error: undefined
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(()=>({ description }));
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(() => ({ amount }));
        }
    }

    onNoteChange = (e) => {
        const note = e.target.value;        
        this.setState(()=>({ note }))
    }

    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(()=>({ createdAt }));
        }        
    }

    onCalendarFocusChange = ({ focused }) => {
        this.setState(() => ({ isCalendarFocused: focused }));
    }

    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            const error = 'Please provide a Description and Amount'
            this.setState(()=>({ error }))
        }else{
            this.setState(() => ({ error:undefined }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100, // convert to a float number of base 10 and convert to cents
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }

    render(){
        return(                                     
            <form className="form" onSubmit={this.onSubmit}>
                { this.state.error && <p className="form__error">{this.state.error}</p> }
                <input
                    className='text-input'
                    type='text'
                    placeholder='Description'
                    autoFocus
                    value={ this.state.description }
                    onChange={ this.onDescriptionChange }
                />
                <input
                    className='text-input' 
                    type='text'
                    placeholder='Amount'
                    value={ this.state.amount }
                    onChange={ this.onAmountChange }
                />
                <SingleDatePicker 
                    date={ this.state.createdAt } 
                    onDateChange={ this.onDateChange } 
                    focused={ this.state.isCalendarFocused } 
                    onFocusChange={ this.onCalendarFocusChange }
                    numberOfMonths = {1}
                    isOutsideRange = {() => false} 
                />
                <textarea
                    className='text-area'
                    placeholder='Enter a note for your expense (Optional)'
                    value={ this.state.note }
                    onChange={ this.onNoteChange }
                >
                </textarea>
                <div>
                    <button className="button">Save Expense</button>
                </div>
                
            </form>            
        )
    }
}


