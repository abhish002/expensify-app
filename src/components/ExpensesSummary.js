import React from 'react';
import { connect } from 'react-redux';
import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';
import { Link } from 'react-router-dom';

export const ExpensesSummary = ({expenseCount, expensesTotal}) => {
    const expenseWord = expenseCount > 1 ? 'expenses' : 'expense';
    const formattedExpenseTotal = numeral(expensesTotal).format('$0,0.00');
    return (
        <div className="page-header">
            <div className="content-container">
            {             
                <h1 className="page-header__title">                    
                    Viewing <span>{expenseCount}</span> {expenseWord} totalling: <span>{formattedExpenseTotal}</span>                    
                </h1>            
            }
            <div className="page-header__actions">
                <Link className="button" to="/create">Add Expense</Link>
            </div>
            </div>            
        </div>
    );
}

const mapStateToProps = (state) => {
    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)/100
    }
}

export default connect(mapStateToProps)(ExpensesSummary);