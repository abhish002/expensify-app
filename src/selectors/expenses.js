import moment from 'moment';

// Get Visible Expenses or Filtered Expenses (filtered results)
export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const momentCreatedAt = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(momentCreatedAt, 'day'): true;
        const endDateMatch = endDate ? endDate.isSameOrBefore(momentCreatedAt, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase()) || expense.note.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if(sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1 : -1
        }else if(sortBy === 'amount'){
            return a.amount < b.amount ? 1 : -1
        }
    });
};