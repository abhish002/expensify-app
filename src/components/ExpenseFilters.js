import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

class ExpenseFilters extends React.Component{
    state = {
        focusedInput: null
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    }

    render(){
        return(
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            className="text-input" 
                            type='text'
                            placeholder='Search Expenses' 
                            value={this.props.filters.text} 
                            onChange={(e)=>{
                                this.props.dispatch(setTextFilter(e.target.value));            
                            }}
                        />
                    </div>
                    <div className="input-group__item">
                        <select className="select" value={this.props.filters.sortBy} onChange={
                            (e) => {                
                                const selectedValue=e.target.value;
                                if(selectedValue==='amount'){
                                    this.props.dispatch(sortByAmount());
                                }else if(selectedValue==='date'){
                                    this.props.dispatch(sortByDate());
                                }                
                            }
                            }
                        >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>                
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker 
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.focusedInput}
                            onFocusChange={focusedInput => this.setState({ focusedInput })}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            showClearDates={true}
                            // isDayHighlighted={()=>true}
                        />
                    </div>
                </div>
            </div>
        )
    }
}

const mapStateToProps = state => {
    return{
        filters: state.filters
    }    
};

export default connect(mapStateToProps)(ExpenseFilters);