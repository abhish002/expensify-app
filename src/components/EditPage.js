import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { editExpense, startRemoveExpense } from '../actions/expenses';

const EditPage = (props) => (
    <div>
        <h1>Edit page.</h1>
        <ExpenseForm 
            expense={props.expense}
            onSubmit={(editProps)=>{                
                props.dispatch(editExpense(props.expense.id, editProps))
                props.history.push('/');
            }}
        />
        <button onClick={()=>{
            const id = props.expense.id;
            props.dispatch(startRemoveExpense({ id }));            
            props.history.push('/');
        }}
        >
        Remove
        </button>        
    </div>
);
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(expense => expense.id === props.match.params.id) 
    }
}

export default connect(mapStateToProps)(EditPage);