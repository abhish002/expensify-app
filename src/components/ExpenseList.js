import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import visibleExpenses from '../selectors/expenses';
import ExpenseFilters from './ExpenseFilters';

export const ExpenseList = (props) => (
    <div>
        <h2>Expenses List</h2>
        <ExpenseFilters />   
        {
            props.expenses.length !== 0 ?                     
            props.expenses.map((expense)=>{
                return (
                    <ExpenseListItem
                        key={expense.id} 
                        {...expense}
                    />
                )
            }) :
            <p>No expenses to list.</p>
        }
    </div>
);

const mapStateToProps = (state)=>{
    return{
        expenses: visibleExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList);