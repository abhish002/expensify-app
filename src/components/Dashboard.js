import React from 'react';
import ExpenseList from '../components/ExpenseList'

const ExpenseDashboardPage = () => (    
    <ExpenseList />
);

export default ExpenseDashboardPage;