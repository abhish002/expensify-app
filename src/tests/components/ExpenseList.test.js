import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

test('should render expense list for provided expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={ expenses }/>)
    expect(wrapper).toMatchSnapshot();
});

test('should render paragraph tag when no expenses are passed', () => {
    const wrapper = shallow(<ExpenseList expenses={ [] }/>)
    expect(wrapper).toMatchSnapshot();
});