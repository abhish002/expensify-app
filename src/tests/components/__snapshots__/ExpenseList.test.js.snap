// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list for provided expenses 1`] = `
<div>
  <h2>
    Expenses List
  </h2>
  <Connect(ExpenseFilters) />
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="book"
    id="1"
    key="1"
    note="book"
  />
  <ExpenseListItem
    amount={109500}
    createdAt={-345600000}
    description="rent"
    id="2"
    key="2"
    note="rent"
  />
  <ExpenseListItem
    amount={4500}
    createdAt={345600000}
    description="gas"
    id="3"
    key="3"
    note="gas"
  />
</div>
`;

exports[`should render paragraph tag when no expenses are passed 1`] = `
<div>
  <h2>
    Expenses List
  </h2>
  <Connect(ExpenseFilters) />
  <p>
    No expenses to list.
  </p>
</div>
`;
